examples/recurrent_attention_machine_translation.py

```diff
@@ -0,0 +1,683 @@
+                for sentence in sentences]
+
+    input_texts_train = get_sentences("train", FROM_LANGUAGE)
+    input_texts_val = get_sentences("val", FROM_LANGUAGE)
+    target_texts_train = get_sentences("train", TO_LANGUAGE)
+    target_texts_val = get_sentences("val", TO_LANGUAGE)
+
+    input_tokenizer = Tokenizer(num_words=MAX_UNIQUE_WORDS)
+    target_tokenizer = Tokenizer(num_words=MAX_UNIQUE_WORDS)
+    input_tokenizer.fit_on_texts(input_texts_train + input_texts_val)
+    target_tokenizer.fit_on_texts(target_texts_train + target_texts_val)
+
+    input_seqs_train = input_tokenizer.texts_to_sequences(input_texts_train)
+    input_seqs_val = input_tokenizer.texts_to_sequences(input_texts_val)
+    target_seqs_train = target_tokenizer.texts_to_sequences(target_texts_train)
+    target_seqs_val = target_tokenizer.texts_to_sequences(target_texts_val)
+
+    input_seqs_train, input_seqs_val, target_seqs_train, target_seqs_val = (
+        pad_sequences(seq, maxlen=MAX_WORDS_PER_SENTENCE, padding='post')
+        for seq in [input_seqs_train,
+                    input_seqs_val,
+                    target_seqs_train,
+                    target_seqs_val]
+    )
+
+    # Build the model
+    x = Input((None,), name="input_sequences")
+    y = Input((None,), name="target_sequences")
+    x_emb = Embedding(input_tokenizer.num_words,
+                      EMBEDDING_SIZE,
+                      mask_zero=True)(x)
+    y_emb = Embedding(target_tokenizer.num_words,
+                      EMBEDDING_SIZE,
+                      mask_zero=True)(y)
+
+    encoder = Bidirectional(GRU(RECURRENT_UNITS, return_sequences=True))
+    x_enc = encoder(x_emb)
+
+    decoder = RNN(
+        cell=DenseAnnotationAttention(
+            cell=GRUCell(RECURRENT_UNITS),
+            units=DENSE_ATTENTION_UNITS),
+        return_sequences=True)
+    h1 = decoder(y_emb, constants=x_enc)
+
+    def dense_maxout(x_):
+        """Implements a dense maxout layer where max is taken
+        over _two_ units"""
+        x_ = Dense(READOUT_HIDDEN_UNITS * 2)(x_)
+        x_1 = x_[:, :READOUT_HIDDEN_UNITS]
+        x_2 = x_[:, READOUT_HIDDEN_UNITS:]
+        return K.max(K.stack([x_1, x_2], axis=-1), axis=-1, keepdims=False)
+
+    h2 = TimeDistributed(Lambda(dense_maxout))(concatenate([h1, y_emb]))
+    y_pred = TimeDistributed(Dense(target_tokenizer.num_words,
+                                   activation='softmax'))(h2)
+
+    model = Model([y, x], y_pred)
+    model.compile(loss='sparse_categorical_crossentropy', optimizer=OPTIMIZER)
+
+    model.fit([target_seqs_train[:, :-1], input_seqs_train],
+              target_seqs_train[:, 1:, None],
+              batch_size=BATCH_SIZE,
+              epochs=EPOCHS,
+              validation_data=(
+                  [target_seqs_val[:, :-1], input_seqs_val],
+                  target_seqs_val[:, 1:, None]))
+
+    # TODO add logic for greedy/beam search generation
```

Comment:
I think stopping the example at `model.fit` is too restrictive, this should be an end-to-end example showing how to do inference as well (like we do in the other translation example).

Reply:
Totally agree. Beam search will make the example even significantly longer more complex (but is most relevant). Will add both greedy and beam search inference to the example and then we can decide.

Reply:
Totally agree. Beam search will make the example even significantly longer more complex (but is most relevant). Will add both greedy and beam search inference to the example and then we can decide.